clippy_utils/src/usage.rs

```diff
@@ -199,3 +199,56 @@ pub fn contains_return_break_continue_macro(expression: &Expr<'_>) -> bool {
     recursive_visitor.visit_expr(expression);
     recursive_visitor.seen_return_break_continue
 }
+
+pub struct UsedAfterExprVisitor<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    expr: &'tcx Expr<'tcx>,
+    definition: HirId,
+    past_self: bool,
+    used_after_self: bool,
+}
+impl<'a, 'tcx> UsedAfterExprVisitor<'a, 'tcx> {
+    pub fn is_found(cx: &'a LateContext<'tcx>, definition: HirId, expr: &'tcx Expr<'_>) -> bool {
+        let mut visitor = UsedAfterExprVisitor {
+            cx,
+            expr,
+            definition,
+            past_self: false,
+            used_after_self: false,
+        };
+        utils::get_enclosing_block(cx, definition).map_or(false, |block| {
+            visitor.visit_block(block);
+            visitor.used_after_self
+        })
+    }
+}
+
+impl<'a, 'tcx> intravisit::Visitor<'tcx> for UsedAfterExprVisitor<'a, 'tcx> {
+    type Map = Map<'tcx>;
+
+    fn nested_visit_map(&mut self) -> NestedVisitorMap<Self::Map> {
+        NestedVisitorMap::OnlyBodies(self.cx.tcx.hir())
+    }
+
+    fn visit_expr(&mut self, expr: &'tcx Expr<'tcx>) {
+        if self.used_after_self {
+            return;
+        }
+
+        if expr.hir_id == self.expr.hir_id {
+            self.past_self = true;
+
+            if let Some(Expr {
+                kind: ExprKind::Loop(..),
+                ..
+            }) = utils::get_enclosing_loop_or_closure(self.cx.tcx, expr)
```

Comment:
I don't think this belongs in this visitor. If loops are a concern, then this function should be called explicitly, instead of hiding it away in this visitor.

Also you don't have to differentiate between loops and closures.